trackpy/masks.py

```diff
@@ -30,19 +35,42 @@ def r_squared_mask(radius, ndim):
         coords = np.array(np.meshgrid(*points, indexing="ij"))
     else:
         coords = np.array([points[0]])
-    r = [(coord/rad)**2 for (coord,rad) in zip(coords,radius)]
+    r = [(coord/rad)**2 for (coord, rad) in zip(coords, radius)]
     r2 = np.sum(coords**2, 0).astype(int)
     r2[sum(r) > 1] = 0
     return r2
+    
+
+@memo
+def x_squared_masks(radius, ndim):
+    "Returns ndim masks with values x^2 inside radius and 0 outside"
+    radius = validate_tuple(radius, ndim)
+    points = [np.arange(-rad, rad + 1) for rad in radius]
+    if len(radius) > 1:
+        coords = np.array(np.meshgrid(*points, indexing="ij"))
+    else:
+        coords = np.array([points[0]])
+    r = [(coord/rad)**2 for (coord, rad) in zip(coords, radius)]
+    masks = np.asarray(coords**2, dtype=int)
+    masks[:, sum(r) > 1] = 0
+    return masks
+
+
+@memo
+def root_sum_x_squared(radius, ndim):
+    "Returns the root of the sum of all x^2 inside the mask for each dim."
+    masks = x_squared_masks(radius, ndim)
+    r2 = np.sum(masks, axis=tuple(range(1, ndim + 1)))  # each ax except first
+    return np.sqrt(r2)
 
 
 @memo
 def theta_mask(radius):
     "Mask of values giving angular position relative to center"
     # 2D only
-    tan_of_coord = lambda y, x: np.arctan2(radius - y, x - radius)
-    diameter = 2*radius + 1
-    return np.fromfunction(tan_of_coord, (diameter, diameter))
+    radius = validate_tuple(radius, 2)
+    tan_of_coord = lambda y, x: np.arctan2(radius[0] - y, x - radius[1])
```

Comment:
The skxray definition is following the numpy definition (
http://docs.scipy.org/doc/numpy/reference/generated/numpy.arctan2.html)
which derives from (x, y) = (r * cos(th), r * sin(y)).

If you invert the y-axis before computing `atan2`   things will not round
trip correctly and computing what pixel a given angle corresponds to will
fail.

On Mon, May 11, 2015 at 10:39 AM Casper van der Wel <
notifications@github.com> wrote:

> In trackpy/masks.py
> <https://github.com/soft-matter/trackpy/pull/239#discussion_r30043883>:
>
> >
> >
> >  @memo
> >  def theta_mask(radius):
> >      "Mask of values giving angular position relative to center"
> >      # 2D only
> > -    tan_of_coord = lambda y, x: np.arctan2(radius - y, x - radius)
> > -    diameter = 2*radius + 1
> > -    return np.fromfunction(tan_of_coord, (diameter, diameter))
> > +    radius = validate_tuple(radius, 2)
> > +    tan_of_coord = lambda y, x: np.arctan2(radius[0] - y, x - radius[1])
>
> I am no mathematician, but I would say that in the upper two quadrants
> (negative y), there should be positive angles and in the lower two
> quadrants (positive y) negative angles.
>
> The scikit-xray reference works the other way around. It measures angles
> counterclockwise *from (x, y) to the positive x-axis*. I think in the
> usual definition it is measured *from the positive x-axis to (x, y)*.
>
> In [6]: angle_grid([1,1], (3,3))
> Out[6]:
> array([[-2.35619449, -1.57079633, -0.78539816],
>        [ 3.14159265,  0.        ,  0.        ],
>        [ 2.35619449,  1.57079633,  0.78539816]])
>
> From a second perspective: the np.arctan2(y, x) is made for normal
> Caretisian coordinates; as our y-axis is flipped, we should do np.arctan2(-y,
> x).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/soft-matter/trackpy/pull/239/files#r30043883>.
>